tiatoolbox/models/dataset/abc.py

```diff
@@ -0,0 +1,152 @@
+# ***** END GPL LICENSE BLOCK *****
+
+
+import torch
+import pathlib
+import numpy as np
+import os
+
+from tiatoolbox.utils.misc import imread
+
+
+class ABCPatchDataset(torch.utils.data.Dataset):
+    """Defines abstract base class for patch dataset.
+
+    Attributes:
+        return_labels (bool, False): `__getitem__` will return both the img and
+        its label. If `label_list` is `None`, `None` is returned
+
+        preproc_func: Preprocessing function used to transform the input data. If
+        supplied, then torch.Compose will be used on the input preproc_list.
+        preproc_list is a list of torchvision transforms for preprocessing the image.
+        The transforms will be applied in the order that they are given in the list.
+        https://pytorch.org/vision/stable/transforms.html.
+
+    """
+
+    def __init__(self, preproc_func=None):
+        super().__init__()
+        self.set_preproc_func(preproc_func)
+        self.data_is_npy_alike = False
+        self.input_list = []
+        self.label_list = []
+
+    def check_input_integrity(self, mode):
+        """Perform check on the input to make sure it is valid."""
+        if mode == "patch":
+            self.data_is_npy_alike = False
+
+            # If input is a list - can contain a list of images or a list of image paths
+            if isinstance(self.input_list, list):
+                is_all_path_list = all(
+                    isinstance(v, (pathlib.Path, str)) for v in self.input_list
+                )
+                is_all_npy_list = all(
+                    isinstance(v, np.ndarray) for v in self.input_list
+                )
+                if not (is_all_path_list or is_all_npy_list):
+                    raise ValueError(
+                        "Input must be either a list/array of images "
+                        "or a list of valid image paths."
+                    )
+
+                shape_list = []
+                # When a list of paths is provided
+                if is_all_path_list:
+                    if any(not os.path.exists(v) for v in self.input_list):
+                        # at least one of the paths are invalid
+                        raise ValueError(
+                            "Input must be either a list/array of images "
+                            "or a list of valid image paths."
+                        )
+                    # Preload test for sanity check
+                    shape_list = [self.load_img(v).shape for v in self.input_list]
+                    self.data_is_npy_alike = False
+                else:
+                    shape_list = [v.shape for v in self.input_list]
+                    self.data_is_npy_alike = True
+
+                if any(len(v) != 3 for v in shape_list):
+                    raise ValueError("Each sample must be an array of the form HWC.")
+
+                max_shape = np.max(shape_list, axis=0)
+                # How will this behave for mixed channel ?
```

Comment:
Is this a question or a comment?

Reply:
My rambling when codeing, removed now. :)